lib/seahorn/Smt/TypeChecker.cc

```diff
@@ -6,41 +6,134 @@
 // local code in this namespace
-
 //==-- main implementation goes here --==//
 class TCVR {
 
+  ExprMap m_cache;
+  std::map<Expr, ExprSet> m_endExpr;
+
+  bool m_isWellFormed;
+  Expr m_errorExp;
+
+  TypeChecker *const m_tc;
+  Expr m_topMost;
+
+  void reset(Expr exp) {
+    if (m_isWellFormed)
+      m_errorExp = Expr();
+
+    m_topMost = Expr();
+    m_isWellFormed = true;
+    m_cache.clear();
+  }
+
+  bool isConst(Expr exp) {
+    return bind::isBoolConst(exp) || bind::isIntConst(exp) ||
+           bind::isRealConst(exp) || bind::isUnintConst(exp) ||
+           bind::isArrayConst(exp) || bind::isStructConst(exp) ||
+           bv::isBvConst(exp);
+  }
+
+  bool isValue(Expr exp) { return isOpX<TRUE>(exp) || isOpX<FALSE>(exp); }
+
+  bool isSimpleType(Expr exp) {
+    return isConst(exp) || isValue(exp) || bv::is_bvnum(exp);
+  }
+
+  void mapEndTypes(Expr exp) {
+    ExprSet set;
+
+    if (exp->arity() == 0) {
+      set.insert(exp);
+    }
+
+    std::for_each(exp->args_begin(), exp->args_end(), [this, &set](Expr child) {
+      if (this->m_endExpr.count(child)) {
+        ExprSet childSet = this->m_endExpr.at(child);
+        set.insert(childSet.begin(), childSet.end());
+      }
+    });
+
+    m_endExpr.insert({exp, set});
+  }
+
   /// Called after children have been visited
   Expr postVisit(Expr exp) {
     LOG("tc", llvm::errs() << "post visiting expression: " << *exp << "\n";);
+
+    Expr type = exp->inferType(exp, *m_tc);
+
+    m_cache.insert({exp, type});
+    mapEndTypes(exp);
+
+    if (isOp<ERROR_TY>(type)) {
+      LOG("tc", llvm::errs()
+                    << "Expression is not well-formed: " << *exp << "\n";);
+
+      if (m_isWellFormed)
+        m_errorExp = exp;
+
+      m_isWellFormed = false;
+    }
+
     return exp;
   }
 
 public:
+  TCVR(TypeChecker *tc) : m_isWellFormed(true), m_tc(tc), m_topMost(Expr()) {}
+
   /// Called before children are visited
   /// Returns false to skip visiting children
   bool preVisit(Expr exp) {
     LOG("tc", llvm::errs() << "pre-visiting: " << *exp << "\n";);
+
+    if (!m_topMost) {
+      LOG("tc", llvm::errs() << "top most expression: " << *exp << "\n";);
+      m_topMost = exp;
+    }
+
+    if (m_cache.count(exp) || !m_isWellFormed) {
+      return false;
+    } else if (isSimpleType(exp)) {
+      m_cache.insert({exp, exp->inferType(exp, *m_tc)});
+
+      ExprSet set;
+      set.insert(exp);
+      m_endExpr.insert({exp, set});
+      return false;
+    }
     return true;
   }
+
   Expr operator()(Expr exp) { return postVisit(exp); }
 
   Expr knownTypeOf(Expr e) {
-    LOG("tc", llvm::errs() << "Request type of: " << *e << "\n";);
-    // if e is non null, return ANY type, ow return nil expression
-    return e ? sort::anyTy(e->efac()) : Expr();
+    Expr knownType;
+
+    if (m_isWellFormed) 
+      knownType = e ? m_cache.at(e) : Expr();
+    else
+      knownType = sort::errorTy(e->efac());
+
+    if (e == m_topMost) // done traversing entire expression
```

Comment:
Not sure why **top most** expression is important. Some comments on that will be useful.

In particular, we would like to run type checker over multiple expressions (think vector of expressions), while re-using the type cache. In this case, top-most expression changes all the time.

Reply:
The problem that I had was that it skips all children and returns an error if it finds something wrong. Without the topmost expression, if you find an error every expression after that will also say that its an error, even if its not.

Reply:
I could keep the top most expression and just not clear the cache when I reset everything

Reply:
or maybe just remove it all together with the assumption that when you find an error it will say error for everything after that?

Reply:
If an expression is typed as `errorTy`, then all of its parents will be typed as `errorTy`. This seems good and also stable (so can be cached). The first expression that was typed as `errorTy` can be logged in a field to be returned to the user. User can reset last error during the next call (or the reset can happen automatically).

Reply:
So with that being said, maybe I'll make the reset happen automatically (using the topmost expression)